Improve gammapy download

Bultako opened the pull request:
This PR contains several improvements and new functionalities for `gammapy download`.
* Code base refactoring
* Tries to solve race condition issues in parallell downloading processes
* Adds MD5 hash checking to skip fetching previously dowloaded datasets
* Adds functionality to download example python scripts
* Adds functionality to download versioned datasets

It addresses issues and comments https://github.com/gammapy/gammapy/issues/1880, https://github.com/gammapy/gammapy/pull/1836#issuecomment-424632483, https://github.com/gammapy/gammapy/commit/e41add2e8ae42a814b74017c28e56402f899e454#r30632763, https://github.com/gammapy/gammapy/issues/172, https://github.com/gammapy/gammapy/pull/1994#issuecomment-456150395

cdeil commented:
I see that you include code to download the Gammapy conda env file via `gammapy download`.
Maybe we should remove that (the `ComputePlan.getenvironment` method)?

The current instructions (https://docs.gammapy.org/0.9/getting-started.html) are to first
```
curl -O https://gammapy.org/download/install/gammapy-0.9-environment.yml
conda env create -f gammapy-0.9-environment.yml
conda activate gammapy-0.9
```
and only then:
```
gammapy download tutorials
cd gammapy-tutorials
export GAMMAPY_DATA=$PWD/datasets
```
So downloading the `gammapy-0.9-environment.yml` again in `gammapy download tutorials` and printing to the console (in `ParallelDownload.show_info`) that the user should create and activate the env again isn't needed and will probably confuse users.

I can see the appeal of putting a copy of `gammapy-0.9-environment.yml` in the `gammapy-tutorials` repo, in case someone wants to recreate the env later. But overall I feel it might be best to remove the handling of `gammapy-0.9-environment.yml` completely from `gammapy download`.

@Bultako - What do you think?

Bultako commented:
@cdeil 

Note that the environment YAML file is fetched **only** when the user asks for the notebooks or scripts linked to a specific stable release **using the `release` flag**. In the case you describe, the environment file will not be downloaded because the `release` flag is not present at any moment in the instructions we provide. The information displayed in the terminal after execution of `gammapy download tutorials` is be the following:

```
*** You might want to declare GAMMAPY_DATA env variable
export GAMMAPY_DATA=/Users/jer/Desktop/gammapy-tutorials/datasets
```

So, I'm +1 to continue downloading the environment YAML file only when a particular release for the tutorials is asked on purpose by the user, so he/she can reproduce the versioned tutorials in a specific conda virtual env. Does this make sense?

cdeil commented:
> So, I'm +1 to continue downloading the environment YAML file only when a particular release for the tutorials is asked on purpose by the user, so he/she can reproduce the versioned tutorials in a specific conda virtual env. Does this make sense?

Yes, that does make sense, we can keep as-is if you like it.

I'm about 50:50 on this, but would probably remove it, because it's not really needed, and not offering it means less and simpler code for `gammapy download`, and also fewer options / documentation for the end user. Note that at the moment it's a hidden feature, not advertised to users. Probably we should add a section "How to Upgrade" at https://docs.gammapy.org/0.9/getting-started.html .
IMO the recommendation should just be to execute the same commands for the different version the users wants, we shouldn't explain the `download --release` option, it will just confuse users which way to use?

@Bultako - Should I do a final review here, or are you still working on this branch?

Bultako commented:
I think version management in general has never been easy for most users, worst for newcomers. `gammapy download` is not the exception, it is complex in its design, it offers quite a collection of features (dealing with versioning in different kinds of material is only one), trying to be highly flexible in the combination of options from the user interface. This complexity in the design reflects on the code, I agree it's not easy to maintain right now. And surely we are lacking of documentation about all this.

But I think version management (the `--release` option) is the feature I would keep the most (you can see that I have added version management for the datasets..) There are other features i.e. fetching specific files, only notebooks or scripts or their combination with the datasets needed, that could be simplified and I'm +1 to do it. (maybe not in this PR?)

> @Bultako - Should I do a final review here, or are you still working on this branch?

You can proceed with the review.

cdeil commented on gammapy/scripts/downloadclasses.py:
```diff
@@ -0,0 +1,316 @@
+
+
+def parse_imagefiles(notebookslist):
+    imagefiles = {}
```
I think using generators ("yield") in the `parse_` functions would save a few lines and be a bit simpler.
Note that if you need lists or dicts instead of generators, you can always call `list` or `dict` on the return value from the generator function. It's not about efficiency in this case, but (minor) simpler / shorter code in the function making the list.

cdeil commented on gammapy/scripts/downloadclasses.py:
```diff
@@ -0,0 +1,316 @@
+                    label = "im: " + im
+                    path = "images/" + im + ".png"
+                    url = str(Path(record["url"]).parent)
+                    url = url.replace(":/", "://")
```
Why `url.replace(":/", "://")`? Is this a data entry bug in the input file? Can it be fixed there and this line removed?

Bultako commented:
@cdeil 
I have addressed all your comments in the review.
I have also made a small refactoring of the code to get something simpler, but I don't think I have really achieved this goal. If you can have a quick look now, with the distance that I do not have, that would be nice. 

There are still the test_*.py  scripts to make so to improve the coverage and other thing to do could be to declare which scripts to deliver with `gammapy download scripts`. At this moment the index file for the scripts is really short. 
https://github.com/gammapy/gammapy/blob/master/examples/scripts.yaml
Maybe we could disable this feature?

cdeil commented:
> Maybe we could disable this feature?

I would just leave it activated, and then in the coming weeks add a useful script or two.

Everything in Gammapy is work in progress, and I would always just default to working towards good v1.0 trying to make progress as fast as possible, instead of spending effort to ship "clean" v0.10 or other intermediate releases.
